File: 프로그래머스/1/1845. 폰켓몬/폰켓몬.py
import random 

def solution(nums):
    answer = 0
    n = len(nums)/2
    
    nums_set = set(nums)
    
    if n > len(nums_set):
        answer = len(nums_set)
    else:
        answer = n
    # nums_sample = random.sample(nums, n)
    
    
    return answer


'''
3 1 2 3 
2마리 선택 가능

n개 중에서 n/2개 선택 가능 
서로 다른 숫자 개수 
n/2마리 중에서 최대한 서로 다른 숫자 많게 고르기

'''

File: Java/프로그래머스/1/1845. 폰켓몬/폰켓몬.java
// 총 N마리 중에서 N/2마리만 선택 가능
// 다양한 종류를 고르기 위해 중복된 종류는 1번만 셈
// 고유한 폰켓몬 종류 수 = Set.size()
// 실제로 고를 수 있는 종류 수 = Math.min(Set.size(), N / 2)

import java.util.*; // 이 패키지 꼭 써줘야 Set과 HashSet 사용 가능 

class Solution {
    public int solution(int[] nums) {
        int answer = 0;
        
        // 중복제거한 요소 
        Set<Integer> set = new HashSet<>();
        for (int n : nums) {
            set.add(n);  // 고유한 종류만 저장됨
        }
        
        // N / 2
        int N = nums.length;
        
        // answer 
        answer = Math.min(set.size(), N / 2);
        
        return answer;
    }
}